File: modulo-govdoc/api/app/api/v1/dev_saas.py
"""Ponte de sessão de desenvolvimento com a plataforma SaaS.

Espelha o `/api/dev/saas` do ChatGov: em ambiente de desenvolvimento o
frontend faz login na API do GovSistem e entrega o `access_token` aqui; o
GovDoc valida a identidade na plataforma (`/auth/me` + `/dashboard`),
provisiona a organização e o usuário just-in-time e devolve um token local.

Sem as flags explícitas (`ENABLE_DEV_SAAS_AUTH` / `ENABLE_DEV_E2E_AUTH`) as
rotas retornam 404 — nunca ativadas no compose de produção.
"""

import logging

import httpx
from fastapi import APIRouter, Depends
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.config import settings
from app.core.database import get_db
from app.core.errors import AppError
from app.core.security import create_access_token, decode_token
from app.models.enums import Profile
from app.models.user import User
from app.schemas.auth import DevSessionRequest
from app.services.provisioning import (
    ensure_institution_provisioned,
    ensure_user_provisioned,
)

logger = logging.getLogger("govdoc.dev_saas")

router = APIRouter(prefix="/auth", tags=["Autenticação (dev)"])


def _rota_disponivel() -> None:
    raise AppError("Rota não encontrada.", 404, "nao_encontrado")


@router.post("/dev/session")
async def dev_session(
    payload: DevSessionRequest,
    db: AsyncSession = Depends(get_db),
):
    if not settings.ENABLE_DEV_SAAS_AUTH:
        _rota_disponivel()

    try:
        decoded = decode_token(payload.access_token)
    except Exception:
        raise AppError("Token do SaaS inválido.", 401, "token_invalido")
    if decoded.get("type") != "access" or not decoded.get("sub"):
        raise AppError("Token do SaaS inválido.", 401, "token_invalido")

    saas_api_url = settings.SAAS_API_URL.rstrip("/")
    if not saas_api_url:
        raise AppError("SAAS_API_URL não configurada.", 503, "integracao_indisponivel")

    headers = {"Authorization": f"Bearer {payload.access_token}"}
    try:
        async with httpx.AsyncClient(timeout=10) as client:
            me_response = await client.get(f"{saas_api_url}/auth/me", headers=headers)
            if not me_response.ok:
                raise AppError("Sessão do GovSistem inválida.", 401, "sessao_invalida")
            user_data = me_response.json()

            if str(user_data.get("id")) != str(decoded.get("sub")):
                raise AppError("Identidade do GovSistem divergente.", 401, "identidade_divergente")

            dashboard_response = await client.get(f"{saas_api_url}/dashboard", headers=headers)
            if not dashboard_response.ok:
                raise AppError(
                    "Não foi possível confirmar os módulos do usuário.",
                    403,
                    "modulos_indisponiveis",
                )
            dashboard = dashboard_response.json()
            modules = dashboard.get("modules") or []
            if not any(str(m.get("slug")) == "govdoc" for m in modules):
                raise AppError(
                    "Usuário sem permissão para acessar o GovDoc.", 403, "modulo_nao_autorizado"
                )
    except AppError:
        raise
    except Exception as exc:
        logger.warning("Falha ao validar sessão no SaaS: %s", exc)
        raise AppError("Não foi possível validar o acesso no GovSistem.", 401, "sessao_invalida")

    organization_id = user_data.get("organization_id") or decoded.get("organization_id")
    if not organization_id:
        raise AppError("Usuário sem órgão vinculado no GovSistem.", 403, "sem_orgao")

    await ensure_institution_provisioned(
        db,
        organization_id=organization_id,
        name=user_data.get("organization_name") or "",
        slug=user_data.get("organization_slug") or "",
    )
    user = await ensure_user_provisioned(
        db,
        user_id=decoded["sub"],
        organization_id=organization_id,
        name=user_data.get("name") or user_data.get("nome") or "",
        email=user_data.get("email") or "",
        roles=decoded.get("roles") or [],
    )
    await db.commit()

    token = create_access_token(
        str(user.id),
        extra={"perfil": user.profile, "email": user.email},
        token_type="module_access",
    )
    return {
        "token": token,
        "usuario": {
            "id": str(user.id),
            "nome": user.name,
            "email": user.email,
            "perfil": user.profile,
            "organization_id": str(organization_id),
        },
    }


@router.post("/dev/e2e-session")
async def e2e_session(db: AsyncSession = Depends(get_db)):
    """Sessão técnica exclusiva dos testes automatizados do ambiente dev."""
    if not (settings.ENABLE_DEV_SAAS_AUTH and settings.ENABLE_DEV_E2E_AUTH):
        _rota_disponivel()

    user = await db.scalar(
        select(User)
        .where(User.profile == Profile.ADMIN_GERAL.value, User.is_active.is_(True))
        .order_by(User.created_at)
        .limit(1)
    )
    if user is None:
        raise AppError("Administrador DEV não encontrado.", 404, "nao_encontrado")

    token = create_access_token(
        str(user.id),
        extra={"perfil": user.profile, "email": user.email},
        token_type="module_access",
    )
    return {"token": token, "usuario": {"id": str(user.id), "nome": user.name}}

File: modulo-chatgov/backend/src/auth/dev-saas.js
import express from 'express';
import bcrypt from 'bcrypt';
import { randomUUID } from 'crypto';
import db from '../db.js';
import { verifyToken, signToken } from './jwt.js';
import { ensureTenantProvisioned } from '../services/provisionamento.js';

const router = express.Router();

// Sessão técnica exclusiva dos testes automatizados do compose DEV.
// Não recebe identidade do cliente, não existe sem a flag e nunca é habilitada
// no compose de produção.
router.post('/e2e-session', async (_req, res) => {
  if (process.env.NODE_ENV !== 'development'
      || (process.env.ENABLE_DEV_E2E_AUTH || '').toLowerCase() !== 'true') {
    return res.status(404).json({ erro: 'Rota não encontrada' });
  }
  const operador = await db.oneOrNone(
    `SELECT o.id, o.nome, o.email, o.papel, o.tenant_id AS "tenantId",
            t.nome AS "tenantNome", t.slug AS "tenantSlug"
     FROM operadores o JOIN tenants t ON t.id = o.tenant_id
     WHERE o.papel = 'admin' AND o.ativo = true
     ORDER BY o.criado_em LIMIT 1`
  );
  if (!operador) return res.status(404).json({ erro: 'Administrador DEV não encontrado' });
  const token = signToken({
    sub: operador.id,
    name: operador.nome,
    email: operador.email,
    papel: operador.papel,
    organization_id: operador.tenantId,
    tenantId: operador.tenantId,
    tenantNome: operador.tenantNome,
    tenantSlug: operador.tenantSlug,
    type: 'module_access',
    module: 'chatgov-dev-e2e',
  });
  return res.json({ token, operador });
});

function papelFromRoles(roles = []) {
  if (roles.some((role) => ['SUPER_ADMIN', 'PLATFORM_ADMIN', 'ADMIN'].includes(role))) {
    return 'admin';
  }
  if (roles.includes('SUPPORT')) return 'supervisor';
  return 'operador';
}

router.post('/session', async (req, res) => {
  if ((process.env.ENABLE_DEV_SAAS_AUTH || '').toLowerCase() !== 'true') {
    return res.status(404).json({ erro: 'Rota não encontrada' });
  }

  const accessToken = req.body?.access_token;
  if (!accessToken) {
    return res.status(400).json({ erro: 'Token do SaaS não fornecido' });
  }

  try {
    const decoded = verifyToken(accessToken);
    if (decoded.type !== 'access' || !decoded.sub) {
      return res.status(401).json({ erro: 'Token do SaaS inválido' });
    }

    const saasApiUrl = (process.env.SAAS_API_URL || '').replace(/\/$/, '');
    if (!saasApiUrl) throw new Error('SAAS_API_URL não configurada');

    const meResponse = await fetch(`${saasApiUrl}/auth/me`, {
      headers: { Authorization: `Bearer ${accessToken}` },
    });
    if (!meResponse.ok) {
      return res.status(401).json({ erro: 'Sessão do GovSistem inválida' });
    }

    const user = await meResponse.json();
    if (String(user.id) !== String(decoded.sub)) {
      return res.status(401).json({ erro: 'Identidade do GovSistem divergente' });
    }

    const dashboardResponse = await fetch(`${saasApiUrl}/dashboard`, {
      headers: { Authorization: `Bearer ${accessToken}` },
    });
    if (!dashboardResponse.ok) {
      return res.status(403).json({ erro: 'Não foi possível confirmar os módulos do usuário' });
    }
    const dashboard = await dashboardResponse.json();
    const canAccessChatGov = Array.isArray(dashboard.modules)
      && dashboard.modules.some((module) => module.slug === 'chatgov');
    if (!canAccessChatGov) {
      return res.status(403).json({ erro: 'Usuário sem permissão para acessar o ChatGov' });
    }

    const organizationId = user.organization_id || decoded.organization_id;
    if (!organizationId) {
      return res.status(403).json({ erro: 'Usuário sem órgão vinculado no GovSistem' });
    }

    await ensureTenantProvisioned({
      organization_id: organizationId,
      name: user.organization_name,
      slug: user.organization_slug,
      is_active: true,
    });

    const roles = Array.isArray(decoded.roles) ? decoded.roles : [];
    const papel = papelFromRoles(roles);
    const email = String(user.email || '').toLowerCase().trim();
    const nome = user.name || user.nome || 'Usuário GovSistem';

    const existing = await db.oneOrNone(
      'SELECT id FROM operadores WHERE id = $1 OR (tenant_id = $2 AND email = $3)',
      [user.id, organizationId, email]
    );

    let operadorId = user.id;
    if (existing) {
      operadorId = existing.id;
      await db.none(
        `UPDATE operadores
         SET tenant_id = $1, nome = $2, email = $3, papel = $4
         WHERE id = $5`,
        [organizationId, nome, email, papel, operadorId]
      );
    } else {
      const placeholderHash = await bcrypt.hash(randomUUID(), 10);
      await db.none(
        `INSERT INTO operadores (id, tenant_id, nome, email, senha_hash, papel)
         VALUES ($1, $2, $3, $4, $5, $6)`,
        [operadorId, organizationId, nome, email, placeholderHash, papel]
      );

      const deptGeral = await db.oneOrNone(
        "SELECT id FROM departamentos WHERE tenant_id = $1 AND nome = 'Geral' AND ativo = true",
        [organizationId]
      );
      if (deptGeral) {
        await db.none(
          `INSERT INTO operador_departamentos (operador_id, departamento_id, tenant_id)
           VALUES ($1, $2, $3) ON CONFLICT DO NOTHING`,
          [operadorId, deptGeral.id, organizationId]
        );
      }
    }

    const token = signToken({
      sub: operadorId,
      name: nome,
      email,
      roles,
      papel,
      organization_id: organizationId,
      tenantId: organizationId,
      tenantNome: user.organization_name || '',
      tenantSlug: user.organization_slug || '',
      type: 'module_access',
      module: 'chatgov-dev',
    });

    return res.json({
      token,
      operador: {
        id: operadorId,
        nome,
        email,
        papel,
        tenantId: organizationId,
        tenantNome: user.organization_name || '',
        tenantSlug: user.organization_slug || '',
      },
    });
  } catch (err) {
    console.error('[Dev SaaS Auth] Erro:', err.message);
    return res.status(401).json({ erro: 'Não foi possível validar o acesso no GovSistem' });
  }
});

export default router;
